File: backend/app/api/disciplinas.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from typing import List

from app.core.database import get_db
from app.core.security import get_current_user, require_role
from app.models.user import User
from app.models.disciplina import Disciplina, DisciplinaProfessor, DisciplinaAluno, ModoEnum
from app.schemas.schemas import DisciplinaOut, DisciplinaCreate

router = APIRouter()


@router.get("/", response_model=List[DisciplinaOut])
async def list_disciplinas(
    db: AsyncSession = Depends(get_db),
    current: User = Depends(get_current_user),
):
    if current.perfil.value == "professor":
        query = (
            select(Disciplina)
            .join(DisciplinaProfessor, DisciplinaProfessor.disciplina_id == Disciplina.id)
            .where(DisciplinaProfessor.professor_id == current.id, Disciplina.ativa == True)
        )
    elif current.perfil.value == "aluno":
        query = (
            select(Disciplina)
            .join(DisciplinaAluno, DisciplinaAluno.disciplina_id == Disciplina.id)
            .where(DisciplinaAluno.aluno_id == current.id, Disciplina.ativa == True)
        )
    else:
        query = select(Disciplina).where(Disciplina.ativa == True)

    result = await db.execute(query.order_by(Disciplina.nome))
    return [DisciplinaOut.model_validate(d) for d in result.scalars().all()]


@router.post("/", response_model=DisciplinaOut)
async def create_disciplina(
    body: DisciplinaCreate,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin")),
):
    disc = Disciplina(nome=body.nome, modo=ModoEnum(body.modo), cor=body.cor)
    db.add(disc)
    await db.commit()
    await db.refresh(disc)
    return DisciplinaOut.model_validate(disc)


@router.post("/{disc_id}/professores/{prof_id}")
async def assign_professor(
    disc_id: int, prof_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin")),
):
    db.add(DisciplinaProfessor(disciplina_id=disc_id, professor_id=prof_id))
    await db.commit()
    return {"ok": True}


@router.post("/{disc_id}/alunos/{aluno_id}")
async def enroll_aluno(
    disc_id: int, aluno_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    existing = await db.execute(
        select(DisciplinaAluno).where(
            DisciplinaAluno.disciplina_id == disc_id,
            DisciplinaAluno.aluno_id == aluno_id,
        )
    )
    if existing.scalar_one_or_none():
        return {"ok": True, "msg": "já matriculado"}
    db.add(DisciplinaAluno(disciplina_id=disc_id, aluno_id=aluno_id))
    await db.commit()
    return {"ok": True}


@router.delete("/{disc_id}/alunos/{aluno_id}")
async def unenroll_aluno(
    disc_id: int, aluno_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    result = await db.execute(
        select(DisciplinaAluno).where(
            DisciplinaAluno.disciplina_id == disc_id,
            DisciplinaAluno.aluno_id == aluno_id,
        )
    )
    da = result.scalar_one_or_none()
    if da:
        await db.delete(da)
        await db.commit()
    return {"ok": True}


@router.get("/{disc_id}/alunos", response_model=List[dict])
async def list_alunos_disciplina(
    disc_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    result = await db.execute(
        select(User)
        .join(DisciplinaAluno, DisciplinaAluno.aluno_id == User.id)
        .where(DisciplinaAluno.disciplina_id == disc_id, User.ativo == True)
        .order_by(User.nome)
    )
    return [{"id": u.id, "nome": u.nome, "matricula": u.matricula} for u in result.scalars().all()]

File: backend/src/routes/disciplinas.js
import express from 'express';
import { prisma } from '../prisma.js';
import { requireAuth, requireRole } from '../middleware/auth.js';

const router = express.Router();

function formatDisciplina(disc) {
  return {
    id: disc.id,
    nome: disc.nome,
    modo: disc.modo,
    cor: disc.cor,
    ativa: disc.ativa,
  };
}

router.get('/', requireAuth, async (req, res) => {
  const current = req.user;
  const perfil = current.perfil;

  if (perfil === 'professor') {
    const disciplinas = await prisma.disciplina.findMany({
      where: {
        ativa: true,
        professores: { some: { professorId: current.id } },
      },
      orderBy: { nome: 'asc' },
    });
    return res.json(disciplinas.map(formatDisciplina));
  }

  if (perfil === 'aluno') {
    const disciplinas = await prisma.disciplina.findMany({
      where: {
        ativa: true,
        alunos: { some: { alunoId: current.id } },
      },
      orderBy: { nome: 'asc' },
    });
    return res.json(disciplinas.map(formatDisciplina));
  }

  const disciplinas = await prisma.disciplina.findMany({
    where: { ativa: true },
    orderBy: { nome: 'asc' },
  });
  return res.json(disciplinas.map(formatDisciplina));
});

router.post('/', requireAuth, requireRole('admin'), async (req, res) => {
  const { nome, modo = 'qr', cor = '#3b82f6' } = req.body;
  if (!nome) {
    return res.status(400).json({ detail: 'Nome da disciplina é obrigatório' });
  }

  const disciplina = await prisma.disciplina.create({
    data: { nome, modo, cor },
  });
  return res.json(formatDisciplina(disciplina));
});

router.post('/:discId/professores/:profId', requireAuth, requireRole('admin'), async (req, res) => {
  const { discId, profId } = req.params;
  await prisma.disciplinaProfessor.create({
    data: {
      disciplinaId: Number(discId),
      professorId: Number(profId),
    },
  }).catch(() => null);
  return res.json({ ok: true });
});

router.post('/:discId/alunos/:alunoId', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const { discId, alunoId } = req.params;
  const existing = await prisma.disciplinaAluno.findFirst({
    where: {
      disciplinaId: Number(discId),
      alunoId: Number(alunoId),
    },
  });
  if (existing) {
    return res.json({ ok: true, msg: 'já matriculado' });
  }
  await prisma.disciplinaAluno.create({
    data: {
      disciplinaId: Number(discId),
      alunoId: Number(alunoId),
    },
  });
  return res.json({ ok: true });
});

router.delete('/:discId/alunos/:alunoId', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const { discId, alunoId } = req.params;
  await prisma.disciplinaAluno.deleteMany({
    where: {
      disciplinaId: Number(discId),
      alunoId: Number(alunoId),
    },
  });
  return res.json({ ok: true });
});

router.get('/:discId/alunos', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const { discId } = req.params;
  const alunos = await prisma.user.findMany({
    where: {
      ativo: true,
      disciplinasAluno: { some: { disciplinaId: Number(discId) } },
    },
    orderBy: { nome: 'asc' },
  });
  return res.json(alunos.map((u) => ({ id: u.id, nome: u.nome, matricula: u.matricula })));
});

export default router;
